src/main/java/hudson/plugins/sshslaves/verifiers/KnownHostsFileKeyVerificationStrategy.java

```diff
@@ -55,64 +52,76 @@
+    if (!(launcher instanceof SSHLauncher)) {
+      return false;
     }
-    
-    @Override
-    public boolean verify(SlaveComputer computer, HostKey hostKey, TaskListener listener) throws Exception {
-        ComputerLauncher launcher = computer.getLauncher();
-        if (!(launcher instanceof SSHLauncher)) {
-            return false;
-        }
-
-        if (!KNOWN_HOSTS_FILE.exists()) {
-            listener.getLogger().println(Messages.KnownHostsFileHostKeyVerifier_NoKnownHostsFile(KNOWN_HOSTS_FILE.getAbsolutePath()));
-            return false;
-        }
-        
-        KnownHosts knownHosts = new KnownHosts(KNOWN_HOSTS_FILE);
-        int result = knownHosts.verifyHostkey(((SSHLauncher)launcher).getHost(), hostKey.getAlgorithm(), hostKey.getKey());
-        
-        if (KnownHosts.HOSTKEY_IS_OK == result) {
-            listener.getLogger().println(Messages.KnownHostsFileHostKeyVerifier_KeyTrusted(SSHLauncher.getTimestamp()));
-            return true;
-        } else if (KnownHosts.HOSTKEY_IS_NEW == result) {
-            listener.getLogger().println(Messages.KnownHostsFileHostKeyVerifier_NewKeyNotTrusted(SSHLauncher.getTimestamp()));
-            return false;
-        } else {
-            listener.getLogger().println(Messages.KnownHostsFileHostKeyVerifier_ChangedKeyNotTrusted(SSHLauncher.getTimestamp()));
-            return false;
-        }
-        
+
+    if (!KNOWN_HOSTS_FILE.exists()) {
+      listener.getLogger().println(Messages.KnownHostsFileHostKeyVerifier_NoKnownHostsFile(KNOWN_HOSTS_FILE.getAbsolutePath()));
+      return false;
     }
 
-    @Override
-    public String[] getPreferredKeyAlgorithms(SlaveComputer computer) throws IOException {
-        ComputerLauncher launcher = computer.getLauncher();
+    SSHLauncher sshLauncher = (SSHLauncher) launcher;
+    String host = sshLauncher.getHost();
+    String hostPort = host + ":" + sshLauncher.getPort();
 
-        if (!(launcher instanceof SSHLauncher) || !KNOWN_HOSTS_FILE.exists()) {
-            return super.getPreferredKeyAlgorithms(computer);
-        }
+    listener.getLogger().println(Messages.KnownHostsFileHostKeyVerifier_SearchingFor(host, KNOWN_HOSTS_FILE));
+    int resultHost = verify(host, hostKey.getAlgorithm(), hostKey.getKey());
 
-        KnownHosts knownHosts = new KnownHosts(KNOWN_HOSTS_FILE);
-        return knownHosts.getPreferredServerHostkeyAlgorithmOrder(((SSHLauncher) launcher).getHost());
+    listener.getLogger().println(Messages.KnownHostsFileHostKeyVerifier_SearchingFor(hostPort, KNOWN_HOSTS_FILE));
+    int resultHostPort = verify(hostPort, hostKey.getAlgorithm(), hostKey.getKey());
+
+    if (KnownHosts.HOSTKEY_IS_OK == resultHost || KnownHosts.HOSTKEY_IS_OK == resultHostPort) {
+      listener.getLogger().println(Messages.KnownHostsFileHostKeyVerifier_KeyTrusted(SSHLauncher.getTimestamp()));
+      return true;
+    } else if (KnownHosts.HOSTKEY_IS_NEW == resultHost && KnownHosts.HOSTKEY_IS_NEW == resultHostPort) {
```

Comment:
Is this and & or an | ? What happens if only one of them is new?

Reply:
if none of both is `HOSTKEY_IS_OK`, it would mean both are HOSTKEY_IS_NEW (so hosts keys are not in known_hosts), or one of them returns an unknown error, so it will report the worst error that happens.

Reply:
More than right, thanks for the clarification!